File: src/Python/pipeline_services/core/metrics_actions.py
from __future__ import annotations

import time
from typing import Any, Dict

from .pipeline import StepControl


def print_metrics(ctx: Any, control: StepControl) -> Any:
    try:
        metrics_map: Dict[str, Any] = {}

        metrics_map["pipeline"] = control.pipeline_name
        metrics_map["shortCircuited"] = control.is_short_circuited()
        metrics_map["errorCount"] = len(control.errors)

        now_nanos = time.perf_counter_ns()
        start_nanos = control.run_start_ns
        pipeline_nanos = 0
        if start_nanos > 0 and now_nanos > start_nanos:
            pipeline_nanos = now_nanos - start_nanos
        metrics_map["pipelineLatencyMs"] = float(pipeline_nanos) / 1_000_000.0

        action_latency_ms: Dict[str, float] = {}
        for timing in control.timings:
            action_latency_ms[timing.action_name] = float(timing.elapsed_nanos) / 1_000_000.0
        metrics_map["actionLatencyMs"] = action_latency_ms

        print(metrics_map)
    except Exception as caught_error:
        print("metricsError=", str(caught_error))
    return ctx

File: src/Go/pipeline_services/core/metrics_actions.go
package core

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

func PrintMetrics[ContextType any](ctx ContextType, control StepControl[ContextType]) (ContextType, error) {
	metricsMap := map[string]any{}
	metricsMap["pipeline"] = control.PipelineName()
	metricsMap["shortCircuited"] = control.IsShortCircuited()
	metricsMap["errorCount"] = len(control.Errors())

	nowNanos := time.Now().UnixNano()
	startNanos := control.RunStartNanos()
	pipelineLatencyNanos := int64(0)
	if startNanos > 0 && nowNanos > startNanos {
		pipelineLatencyNanos = nowNanos - startNanos
	}
	metricsMap["pipelineLatencyMs"] = float64(pipelineLatencyNanos) / 1_000_000.0

	actionLatencyMs := map[string]float64{}
	actionTimings := control.ActionTimings()
	for index := 0; index < len(actionTimings); index++ {
		timing := actionTimings[index]
		actionLatencyMs[timing.ActionName] = float64(timing.ElapsedNanos) / 1_000_000.0
	}
	metricsMap["actionLatencyMs"] = actionLatencyMs

	jsonBytes, marshalError := json.Marshal(metricsMap)
	if marshalError != nil {
		fmt.Fprintf(os.Stdout, "metricsError=%v\n", marshalError)
		return ctx, nil
	}

	fmt.Fprintf(os.Stdout, "%s\n", string(jsonBytes))
	return ctx, nil
}
